### PR TITLE
Fixes NGRAPH-1191 - BN refactor and inference

### PR DESCRIPTION
**Pull request by cconvey**
* Refactors emitter code for MXNet's Batchnorm operator.

* Refactored code handles traditional training and inference versions of
  the operator.

* No support currently for MXNet's hybrid-mode Batchnorm (like trainig
  mode, but the mean and variance are supplied by the caller rather than
  computed internally by the BN operation.) If such an operator is
  encountered, an error is thrown.

* In cases where nGraph is known to handle the Batchnorm operation
  with an efficient kernel, nGraph's BatchNorm operator is used.
  In all other cases batchnorm is computed using a collection of
  other nGraph operators.

## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/ngraph/ops/batchnorm.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include "ngraph_emitter.h"
+
+namespace ngraph_bridge {
+
+// Create an subgraph of nGraph nodes that's functionally equivalent to nGraph's training-version
+// BatchNorm op, but which doesn't actually use that particular nGraph op.
+// This function exists as a crutch until nGraph's BatchNorm op is fully implemented.
+std::tuple<NgraphNodePtr, NgraphNodePtr, NgraphNodePtr>
+create_batchnorm_training_without_ngraph_bn_op(
+    const float epsilon,
```
Comment (mbrookhart) [inline src/ngraph/ops/batchnorm.h]: Can we use a string instead of a float? I don't want to lose precision on multiple data types.

**Comment by cconvey** [comment]
TESTNOW

File: src/ngraph/ngraph_emitter.cc [hunk 2]
```diff
@@ -923,112 +925,99 @@ void Emitter::CreateLayerOps() {
...
-        ng_in_gamma, convert_order, convert_shape);
-    ng_in_beta = std::make_shared<ngraph::op::Reshape>(
-        ng_in_beta, convert_order, convert_shape);
+    //----------------------------------------------------------------------------------------------
+    // Hybrid mode: use externally supplied mean/variance (as with inference), but also allow
+    // autodifferentiation (as with training).
+    //----------------------------------------------------------------------------------------------
+    if ((exe_mode_ == GraphExeMode::kTrain) && (use_global_stats)) {
+      CHECK(false && "Hybrid-mode batchnorm not implemented yet.");
+      return {};
```
Comment (mbrookhart) [inline src/ngraph/ngraph_emitter.cc]: I think we should use `create_batchnorm_inference_without_ngraph_bn_op` and autodiff it?

File: src/ngraph/ngraph_emitter.cc [hunk 3]
```diff
@@ -121,8 +123,8 @@ NgraphNodePtr Emitter::ReduceAxes(
 NgraphNodePtr Emitter::ReduceAxes(
     const NodePtr& node,
     const std::function<NgraphNodePtr(const NgraphNodePtr&,
-                                      const ngraph::AxisSet&)>& func) {
-  auto input = op_map_[node->inputs_[0]];
+                                      const ngraph::AxisSet&)>& func) const {
+  auto input = checked_lookup(op_map_, node->inputs_[0]);
```
Comment (mbrookhart) [inline src/ngraph/ngraph_emitter.cc]: Why not .at()? This seems unnecessary for a safe lookup.

File: src/ngraph/ops/batchnorm.cc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+  const ngraph::Shape channel_vector_plus_axes_shape = get_vector_plus_axes_shape(
+      batch_data_rank, channel_axis, channel_axis_length);
+
+  const ngraph::element::Type et = ng_in_data->get_element_type();
+  CHECK(ng_beta->get_element_type() == et);
+
+  const NgraphNodePtr ng_epsilon_shaped = makeConstant(et, channel_vector_plus_axes_shape, epsilon);
+
+  const NgraphNodePtr denom = std::make_shared<ngraph::op::Sqrt>(ng_var_shaped + ng_epsilon_shaped);
```
Comment (mbrookhart) [inline src/ngraph/ops/batchnorm.cc]: Why don't you use create_batchnorm_basic_computation_nodes here?

File: src/ngraph/ngraph_utils.h [hunk 5]
```diff
@@ -271,6 +271,50 @@ inline std::size_t hash_combine(const std::size_t& seed, const T& val) {
 void dump_graph(std::shared_ptr<ngraph::Function> f, std::string src_loc = "",
                 std::string filename_suffix = "");
 
+// We define the term "vector plus axes" to refer to tensor shapes meeting the following criteria:
```
Comment (mbrookhart) [inline src/ngraph/ngraph_utils.h]: Nitpick: "vector plus axes" takes a while to digest. Would vector_with_expanded_dims make more sense? 

I get what you're doing here, and it's good to abstract it, I just find the naming a little confusing. Naming, it's hard.

File: src/ngraph/ngraph_emitter.cc (hunk 2 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_emitter.cc]: I thought @pruthviIntel put in a PR for doing training like this with the op?

Comment (cconvey) [reply src/ngraph/ngraph_emitter.cc]: Yeah, I've already worked out how to do this with and without nGraphs' BN op.  But none of our test cases seems to exercise this, and I wanted to get the PR out ASAP since it's been a while.

Comment (mbrookhart) [reply src/ngraph/ngraph_emitter.cc]: It's tested in tests/python/gpu/test_operator_gpu.py

File: src/ngraph/ops/batchnorm.h (hunk 1 above)
Comment (cconvey) [reply src/ngraph/ops/batchnorm.h]: Do we expect any values for epsilon that require more precision than `float`?  There's a few reasons I'm doubtful:
- The default value is `0.001f`.
- The extra precision would only be meaningful in graphs that process data more precise than f32 (which I guess just leaves f64).

f32 is all we care about immediately, if our MKL-DNN kernel support in nGraph's BN op is any indication.

And if we need something more precise, I would think we'd want `const double epsilon` here, rather than a string.

File: src/ngraph/ngraph_emitter.cc (hunk 3 above)
Comment (cconvey) [reply src/ngraph/ngraph_emitter.cc]: What i wanted here was the ability to make the method `const`.  

`operator[]` on `std::map` is not `const`, because it will create an element if one doesn't exist.  So I created `checked_lookup` as a convenience method to do the lookup without the possibility of modifying the map.

I defined `checked_lookup(...)` to fail loudly (exception, I think) it the lookup key is not present in the map, for several reasons:
- I wanted to keep its signature simple, s.t. the return value is the key's dependent value.
- I didn't want to introduce a "maybe" or "optional" type to represent the possibility of a failed lookup, because in this context a failed lookup really would be a surprise.

File: src/ngraph/ops/batchnorm.cc (hunk 4 above)
Comment (cconvey) [reply src/ngraph/ops/batchnorm.cc]: Nice catch.  Missed it due to editing on a small screen on a bumpy train :)  Will fix.

File: src/ngraph/ops/batchnorm.h (hunk 1 above)
Comment (mbrookhart) [reply src/ngraph/ops/batchnorm.h]: It's stored as a string in mxnet, so using a string is preserving whatever information is in there already.

Not a huge deal, f32 is probably good enough.

File: src/ngraph/ngraph_emitter.cc (hunk 3 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_emitter.cc]: Does at() not meet these requirements? Just curious, not against the utility.

File: src/ngraph/ngraph_utils.h (hunk 5 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_utils.h]: Sounds good. Honestly not sure what's best, just reporting my initial confusion.

File: src/ngraph/ngraph_emitter.cc (hunk 3 above)
Comment (cconvey) [reply src/ngraph/ngraph_emitter.cc]: Actually yeah, it does. Good catch.  I thought you were asking about using `at()` on the vector that's present on that line.  Will fix.

File: src/ngraph/ngraph_emitter.cc (hunk 2 above)
Comment (cconvey) [reply src/ngraph/ngraph_emitter.cc]: > It's tested in tests/python/gpu/test_operator_gpu.py

When I ran that the other day, I didn't see my handler for this case get tickled.  Will double check.
